File: apps/backend/src/common/role-hierarchy.ts
import { RoleName } from '@prisma/client';

/** Roles are cumulative: each level includes everything the ones before it can do. */
export const ROLE_RANK: Record<RoleName, number> = {
  ATTENDANT: 0,
  SUPERVISOR: 1,
  ADMINISTRATOR: 2,
  OWNER: 3,
};

export function roleAtLeast(actual: RoleName, required: RoleName): boolean {
  return ROLE_RANK[actual] >= ROLE_RANK[required];
}

/** True if `actual` meets or exceeds the lowest of the acceptable roles. */
export function roleSatisfiesAny(actual: RoleName, acceptable: RoleName[]): boolean {
  if (acceptable.length === 0) return true;
  const minRequired = Math.min(...acceptable.map((r) => ROLE_RANK[r]));
  return ROLE_RANK[actual] >= minRequired;
}

File: apps/mobile/lib/core/role_hierarchy.dart
const _roleRank = {
  'ATTENDANT': 0,
  'SUPERVISOR': 1,
  'ADMINISTRATOR': 2,
  'OWNER': 3,
};

/// True if [actual]'s role outranks or matches [required] in the
/// ATTENDANT < SUPERVISOR < ADMINISTRATOR < OWNER hierarchy — e.g.
/// `roleAtLeast('OWNER', 'SUPERVISOR')` is true.
bool roleAtLeast(String actual, String required) => (_roleRank[actual] ?? -1) >= (_roleRank[required] ?? 999);
